include/pybind11/stl.h

```diff
@@ -251,12 +251,7 @@ struct array_caster {
         return l.release();
     }
 
-    PYBIND11_TYPE_CASTER(ArrayType,
-                         const_name("List[") + value_conv::name
-                             + const_name<Resizable>(const_name(""),
-                                                     const_name("[") + const_name<Size>()
-                                                         + const_name("]"))
-                             + const_name("]"));
+    PYBIND11_TYPE_CASTER(ArrayType, const_name("List[") + value_conv::name + const_name("]"));
```

Comment:
I think if you make the string `typing.Annotated[List[type], size]` then pybind11-stubgen will generate the correct thing from this (and if it doesn't then you can just modify it to do so).

Reply:
Also note, it's just a string. You don't need typing extensions in pybind11, that's the responsibility of whoever interprets the string.

Reply:
`Annotated[List[type], size]` appears to be the intended standard way to hint the length of a list. [PEP 593](https://peps.python.org/pep-0593/) shows this use case as an example. 

I would just remove `typing.` from your suggestion since `List` is not prefixed either.

> This PR achieves MyPy compliance in exchange for potentially very confusing situations where a pybind11-bound function does not accept an input array, and the error message provides zero context why that is (input has the wrong size)

@wjakob My understanding is that this PR only changes the type hint and does not affect the error message otherwise. Is there a situation where `Annotated[List[str], 2]` would be inferior to `List[str[2]]`?